src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenRefReadonlyParameterTests.cs

```diff
@@ -0,0 +1,566 @@
+            		.custom instance void System.Runtime.CompilerServices.RequiresLocationAttribute::.ctor() = (
+            			01 00 00 00
+            		)
+            	// Method begins at RVA 0x2087
+            	// Code size 1 (0x1)
+            	.maxstack 8
+            	IL_0000: ret
+            } // end of method Program::'<<Main>$>g__local|0_0'
+            """);
+    }
+
+    [Theory]
+    [InlineData("var x = (ref readonly int p) => { };")]
+    [InlineData("var x = local; void local(ref readonly int p) { }")]
+    public void AnonymousDelegate(string def)
+    {
+        var source = $"""
+            {def}
+            System.Console.WriteLine(((object)x).GetType());
+            """;
+        var verifier = CompileAndVerify(source, targetFramework: TargetFramework.NetStandard20,
+            sourceSymbolValidator: VerifyRequiresLocationAttributeSynthesized, symbolValidator: VerifyRequiresLocationAttributeSynthesized,
+            expectedOutput: "<>f__AnonymousDelegate0`1[System.Int32]");
+        verifier.VerifyDiagnostics();
+        verifier.VerifyMethodIL("<>f__AnonymousDelegate0`1", "Invoke", """
+            .method public hidebysig newslot virtual 
+            	instance void Invoke (
+            		[in] !T1& arg
+            	) runtime managed 
+            {
+            	.param [1]
+            		.custom instance void System.Runtime.CompilerServices.RequiresLocationAttribute::.ctor() = (
+            			01 00 00 00
+            		)
+            } // end of method '<>f__AnonymousDelegate0`1'::Invoke
+            """);
+    }
+
+    [Fact]
+    public void FunctionPointer()
+    {
+        var source = """
+            class C
+            {
+                public unsafe void M(delegate*<ref readonly int, void> p) { }
+            }
+            """;
+        var verifier = CompileAndVerify(source, options: TestOptions.UnsafeReleaseDll, targetFramework: TargetFramework.NetStandard20,
+            sourceSymbolValidator: verify, symbolValidator: verifyMetadata);
+        verifier.VerifyDiagnostics();
+        verifier.VerifyMethodIL("C", "M", """
+            .method public hidebysig 
+            	instance void M (
+            		method void *(int32& modreq([netstandard]System.Runtime.InteropServices.InAttribute)) p
+            	) cil managed 
+            {
+            	// Method begins at RVA 0x2067
+            	// Code size 1 (0x1)
+            	.maxstack 8
+            	IL_0000: ret
+            } // end of method C::M
+            """);
+
+        static void verify(ModuleSymbol m)
+        {
+            var p = m.GlobalNamespace.GetMember<MethodSymbol>("C.M").Parameters.Single();
+            var ptr = (FunctionPointerTypeSymbol)p.Type;
+            Assert.Equal(RefKind.RefReadOnlyParameter, ptr.Signature.Parameters.Single().RefKind);
+            Assert.Null(m.GlobalNamespace.GetMember<NamedTypeSymbol>(RequiresLocationAttributeQualifiedName));
+        }
+
+        static void verifyMetadata(ModuleSymbol m)
```

Comment:
Thanks.
FWIW, there's a common pattern for such verification, using a single local function and using `bool inSource = module is SourceModuleSymbol;`. It's useful when the source and the metadata cases are mostly similar (which they usually are).